### DISCUSSION TITLE
Unexpected nodes generated in model

### DISCUSSION THREAD

**Discussion by eort** [lnccbrown/HSSM#298] [Q&A] [answered]
Hi all, 

I am trying to set up a hierarchical regression on multiple DDM parameters. In my last attempt, I came across a surprising collection of nodes when printing out arviz.summary on the model inference data. 
Here is the model I set up: 

```
model = hssm.HSSM(data=data, z=0.5, model='ddm', 
    loglik_kind='approx_differentiable',  extra_namespace={"lvl": cfg['drugs']}, hierarchical=True,
    include=[
                    {
                        "name": "v",
                        "formula":"v ~ 1 + (1|subj_idx) + coh*C(drug, levels=lvl)",
                        "link": "identity"
                    },
                    {
                        "name": "a",
                        "formula":"a ~ C(drug, levels=lvl)",
                        "link": "identity"
                    }

    ])
```
with this being the associated graph:
![image](https://github.com/lnccbrown/HSSM/files/12893110/model_graph.pdf)
After the sampling, I checked the summary with arviz and got a table that has following row indices:

![image](https://github.com/lnccbrown/HSSM/assets/8819465/86d90bb3-dae0-49e8-a0e6-27f8e8272283)
Critically, there are 60 of `v_1|subj_idx[XX]` nodes, which matches the number of subjects I have. However, for a it generated 90563  a[XXXXX] nodes, which is the number of trials I have in the dataset, which is not what I want. 
It seems that by using the `hierarchical` flag, and not specifying which variable in the dataset to use as random factor, hssm uses all row index? Not sure whether this is intentional, but it appears to be a little odd. So, if you wanted a random effect only on one parameter, but not on another, that wouldn't be possible? 

Thanks, 
Eduard

**Comment by frankmj** [comment, marked as answer]
Hi Eduard,

this is not actually an error in the model estimation -- it is not
estimating 90K different parameters for a -- but rather it simply stores
the computed values for each regressor on each observation once the
regressors from that trial are taken into account (ie. the
deterministic outputs of the regression). These are stored in the trace
object so arviz displays them by default. You can simply pass an argument
to arviz to exclude those outputs using e.g. var_names = ['~a']



 Michael
Michael J Frank, PhD | Edgar L. Marston Professor
Director, Carney Center for Computational Brain Science
<https://www.brown.edu/carney/ccbs>
Laboratory of Neural Computation and Cognition <https://www.lnccbrown.com/>

Brown University
website <http://ski.clps.brown.edu>

**Reply by eort** [reply]
Ah ok, I see. But are those 60 subject wise v-nodes separate parameters? I guess I am bit confused why these two parameters (a and v) are treated differently then. 
Is there a way to only select the fitted nodes so that I can check convergence, etc.? Or do I need to select them individually per model, based on what I specified?
Thanks in any case!

**Comment by frankmj** [comment]
THere are only 60 nodes v|subject because these are subject-wise
intercepts. Technically there should also be deterministic v nodes for each
trial based on your regression as well, separately from the subject wise
intercepts. You should be able to look at r-hat for all the parameters that
you want to look at by using var_names (to either include selected ones or
exclude others). I think the development team is working on a way to try to
make this less cumbersome so that it doesn't display the deterministics by
default.

**Reply by eort** [reply]
Alright, thanks again!

**Comment by digicosmos86** [comment]
Additionally, the `hierarchical` knob here is intended to convert the entire model hierarchical. When it's set to `True`, every unspecified parameter will essentially have a random intercept according to `participant_id` column, but it will do nothing if `participant_id` column is not present in the data. Therefore, `hierarchical` setting is not doing anything here. We have give this some thought and will throw an error in the next version if `participant_id` is not present in the next version.

Please see the documentation [here](https://lnccbrown.github.io/HSSM/api/hssm/#hssm.HSSM) for more details

**Reply by eort** [reply]
Yes, thanks. I saw that. I tried later to replace my `subj_idx` with `participant_id`, and indeed random intercepts are added to all three parameters (a, t, v). Interestingly though, those deterministics only appear for t and a, but never for v.

**Reply by digicosmos86** [reply]
Yeah this is a bambi-specific behavior. We have communicated this with the bambi developers and they are working on a fix. Please stay tuned :)
